### PR TITLE
Add ova crate

### LINKED ISSUE #129: [Decider] Create Basic Deciders
Create basic deciders

- [x] And decider
- [ ] Not decider
- [x] Witness exists decider
- [ ] ForAllSuchThat decider

### PR DESCRIPTION
**Pull request by syuhei176**
### Deciders

- [x] ForAllSuchThatDecider
- [x] AndDecider
- [x] NotDecider
- [x] PreimageExistsDecider
- [x] SignedByDecider

### Quantifiers

- [x] IntegerRangeQuantifier
- [x] NonnegativeIntegerLessThanQuantifier

### Specs

https://github.com/cryptoeconomicslab/plasma-rust-framework/pull/142/files#diff-6fd1ca045cc7d855bedda67a8b02b05aR18-R48

Close #129

### CODE REVIEW

**Comment by syuhei176** [comment]
One difference from PG is we have PropertyExecuter to traverse Property tree.

**Comment by syuhei176** [comment]
I will do #135 and, Plasma and Channel stuff in next PR!

File: ovm/src/property_executer.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+use crate::error::Error;
+use crate::quantifiers::{IntegerRangeQuantifier, NonnegativeIntegerLessThanQuantifier};
+use crate::types::Decider;
+use crate::types::{Decision, Property, Quantifier, QuantifierResult};
+use bytes::Bytes;
+use plasma_db::impls::kvs::CoreDbLevelDbImpl;
+use plasma_db::traits::db::DatabaseTrait;
+
+/// Core runtime for Property
+pub struct PropertyExecuter {
```
Comment (tkmct) [inline ovm/src/property_executer.rs]: I think 'Executor' is the right spell.

File: ovm/src/property_executer.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    db: CoreDbLevelDbImpl,
+}
+
+impl Default for PropertyExecuter {
+    fn default() -> Self {
+        PropertyExecuter {
+            db: CoreDbLevelDbImpl::open("test"),
```
Comment (tkmct) [inline ovm/src/property_executer.rs]: Properties will be stored in this DB?
Or decisions are stored?

File: ovm/src/property_executer.rs (hunk 1 above)
Comment (syuhei176) [reply ovm/src/property_executer.rs]: Thanks! I'll fix it.

File: ovm/src/deciders/not_decider.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+impl Decider for NotDecider {
+    type Input = NotDeciderInput;
+    fn decide(
+        decider: &PropertyExecuter,
+        input: &NotDeciderInput,
+        _witness: Option<&Bytes>,
+    ) -> Result<Decision, Error> {
+        let decision = decider.decide(input.get_property(), Some(input.get_witness()))?;
```
Comment (tkmct) [inline ovm/src/deciders/not_decider.rs]: I'm not sure why we need to `decider.decide` inside Decider's decide function.

Comment (tkmct) [reply ovm/src/deciders/not_decider.rs]: Maybe it's just naming problem? I'm not sure.

File: ovm/src/property_executer.rs (hunk 2 above)
Comment (syuhei176) [reply ovm/src/property_executer.rs]: Witness is stored as DecisionValue in this DB.
https://github.com/cryptoeconomicslab/plasma-rust-framework/pull/142/files#diff-f6bdfefbe6b1e02b058aeb7acec9fcefR68

File: ovm/src/deciders/not_decider.rs (hunk 3 above)
Comment (syuhei176) [reply ovm/src/deciders/not_decider.rs]: This is the way to decide nested property in PropertyExecutor.
https://github.com/plasma-group/pigi/blob/master/packages/core/src/app/ovm/deciders/not-decider.ts#L22

File: ovm/src/property_executer.rs (hunk 2 above)
Comment (tkmct) [reply ovm/src/property_executer.rs]: Ah, so KVSDecider in pigi is handled in this way in Rust?

Comment (syuhei176) [reply ovm/src/property_executer.rs]: Yes, right! I'm not sure which implementation is better but only PropertyExecutor has persistent storage is more clear.

File: ovm/src/deciders/not_decider.rs (hunk 3 above)
Comment (tkmct) [reply ovm/src/deciders/not_decider.rs]: I see! Now I understand what 'property tree traverse' means.
Can't we write something like `input.get_property().get_decider().decide(input)`?

Comment (syuhei176) [reply ovm/src/deciders/not_decider.rs]: hmm. I don't know how to add `get_decider` to `Property` now. Deciders have different interfaces and these are determined in dynamic. Need more Rust skill...

Comment (syuhei176) [reply ovm/src/deciders/not_decider.rs]: I think we can use mixin style.

```
pub trait DecideMixin {
    fn decide(&self, decider: &PropertyExecutor, witness: &Bytes) -> Decision;
}
impl DecideMixin for Property {
    fn decide(&self, decider: &PropertyExecutor, witness: &Bytes) -> Decision {
        decider.decide(self, witness)
        println!("foo {}", self);
    }
}

let decision = input.get_property().decide(decider, Some(input.get_witness()))?;
 
```

Comment (tkmct) [reply ovm/src/deciders/not_decider.rs]: Looks good!

File: ovm/src/property_executer.rs (hunk 1 above)
Comment (syuhei176) [reply ovm/src/property_executer.rs]: Fixed!

File: ovm/src/deciders/not_decider.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+impl Decider for NotDecider {
+    type Input = NotDeciderInput;
+    fn decide(
+        decider: &PropertyExecutor,
+        input: &NotDeciderInput,
+        _witness: Option<&Bytes>,
+    ) -> Result<Decision, Error> {
+        let decision = input
+            .get_property()
+            .decide(decider, Some(input.get_witness()))?;
```
Comment (syuhei176) [inline ovm/src/deciders/not_decider.rs]: Note: unlike PG's implementation, we use Property to decide. (In PG's implementation, `property.decide(property.input, witness)` )

File: ovm/src/deciders/not_decider.rs (hunk 3 above)
Comment (syuhei176) [reply ovm/src/deciders/not_decider.rs]: I added DecideMixin!

**Comment by tkmct** [comment]
LGTM!

**Pull request merged**